include/pybind11/pybind11.h

```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
     return make_iterator(std::begin(value), std::end(value), extra...);
 }
 
+NAMESPACE_BEGIN(detail)
+// These have to be separated like this because the lambda in the proper version won't compile if
+// the C++ types aren't implicitly convertible.
+template <typename InputType, typename OutputType>
+typename std::enable_if<std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &input_type) {
+    // Some validity checks on output type:
+    if (!std::is_destructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not destructible");
+    if (!std::is_move_constructible<OutputType>::value && !std::is_copy_constructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not move- or copy-constructible");
+
+    input_type.implicit_cpp_conversions.emplace(std::type_index(typeid(OutputType)),
+            [](void *input) -> void * {
+                // Takes a pointer to the InputType object, allocates an OutputType with the implicit
+                // conversion value and moves/copies the converted value into it, returning a pointer.
+                // Returns nullptr if input is nullptr.  The caller is responsible for destruction.
+                if (!input) return nullptr;
+                OutputType t = *reinterpret_cast<InputType*>(input);
```

Comment:
Why the extra move -- isn't this enough? ``return new OutputType(*reinterpret_cast<InputType*>(input));``